runtime-sdk/modules/evm/src/precompile/confidential.rs

```diff
@@ -1,21 +1,71 @@
-use ethabi::ParamType;
+use std::{collections::HashMap, convert::TryInto, sync::Mutex};
+
+use ethabi::{ParamType, Token};
 use evm::{
     executor::stack::{PrecompileFailure, PrecompileOutput},
     Context, ExitError, ExitRevert, ExitSucceed,
 };
 use hmac::{Hmac, Mac, NewMac as _};
-use oasis_runtime_sdk::core::common::crypto::mrae::deoxysii::{DeoxysII, KEY_SIZE, NONCE_SIZE};
+
+use once_cell::sync::Lazy;
+
+use oasis_runtime_sdk::{
+    core::common::crypto::mrae::deoxysii::{DeoxysII, KEY_SIZE, NONCE_SIZE},
+    crypto::signature::{self, SignatureType},
+};
 
 use crate::backend::{EVMBackendExt, RNG_MAX_BYTES};
 
 use super::{linear_cost, multilinear_cost, PrecompileResult};
 
+/// Length of an EVM word, in bytes.
+pub const WORD: usize = 32;
+
+/// The base cost for x25519 key derivation.
+const X25519_KEY_DERIVATION_BASE_COST: u64 = 100_000;
+
 /// The base setup cost for encryption and decryption.
 const DEOXYSII_BASE_COST: u64 = 50_000;
 /// The cost for encryption and decryption per word of input.
 const DEOXYSII_WORD_COST: u64 = 100;
-/// Length of an EVM word, in bytes.
-const WORD: usize = 32;
+
+/// The cost of a key pair generation operation, per method.
+static KEYPAIR_GENERATE_BASE_COST: Lazy<Mutex<HashMap<SignatureType, u64>>> = Lazy::new(|| {
+    let mut costs = HashMap::with_capacity(6);
```

Comment:
You could use `HashMap::from([(SignatureType::Ed25519_Oasis, 35_000), ( ... ), ...])` to avoid some code.